Support injected content in Web Extensions.

xeenon opened the pull request:
#### 5b5387abfac2be339ec93a67d9e8db023aa52b61
<pre>
Support injected content in Web Extensions.
<a href="https://bugs.webkit.org/show_bug.cgi?id=246486">https://bugs.webkit.org/show_bug.cgi?id=246486</a>

Reviewed by Brian Weinstein.

* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtension.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtension.mm:
(-[_WKWebExtension _hasStaticInjectedContentForURL:]): Renamed from hasInjectedContentForURL:.
(-[_WKWebExtension hasInjectedContentForURL:]): Deleted.
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionContext.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionContext.mm:
(-[_WKWebExtensionContext hasInjectedContentForURL:]): Added.
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionController.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionPrivate.h:
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionCocoa.mm:
(WebKit::WebExtension::removeError): Fix lambda capture to capture only this.
(WebKit::WebExtension::recordError): Ditto.
(WebKit::WebExtension::resourceStringForPath): Added. Helper for string resources.
(WebKit::WebExtension::resourceDataForPath): Don&apos;t replace cached strings.
(WebKit::WebExtension::staticInjectedContents): Renamed from injectedContents().
(WebKit::WebExtension::hasStaticInjectedContentForURL): Renamed from hasInjectedContentForURL().
(WebKit::WebExtension::populateContentScriptPropertiesIfNeeded): Use m_staticInjectedContents.
(WebKit::WebExtension::allRequestedMatchPatterns): Ditto.
(WebKit::WebExtension::injectedContents): Deleted.
(WebKit::WebExtension::hasInjectedContentForURL): Deleted.
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
(WebKit::WebExtensionContext::load): Call addInjectedContent(). Adjust FIXMEs.
(WebKit::WebExtensionContext::unload): Call removeInjectedContent(). Move nil setting to end.
(WebKit::WebExtensionContext::injectedContents): Added with FIXME for dynamic scripts.
(WebKit::WebExtensionContext::hasInjectedContentForURL): Added. This is better here than WebExtension
since dynamic scripts will make the answer wrong on WebExtension but correct here.
(WebKit::WebExtensionContext::postAsyncNotification): Fix lambda capture to fix crashes.
(WebKit::WebExtensionContext::grantPermissionMatchPatterns): Call addInjectedContent().
(WebKit::WebExtensionContext::denyPermissionMatchPatterns): Call updateInjectedContent().
(WebKit::WebExtensionContext::removeGrantedPermissionMatchPatterns): Call removeInjectedContent().
(WebKit::WebExtensionContext::removeDeniedPermissionMatchPatterns): Call updateInjectedContent().
(WebKit::WebExtensionContext::loadBackgroundWebView): Fix lambda capture to capture only this.
(WebKit::WebExtensionContext::addInjectedContent): Added.
(WebKit::toImpl): Added.
(WebKit::WebExtensionContext::removeInjectedContent): Added.
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionControllerCocoa.mm:
(WebKit::WebExtensionController::load): Fix the loading order again to fix some ASSERTs.
(WebKit::WebExtensionController::addPage): Call new addProcessPool() and addUserContentController() helpers.
(WebKit::WebExtensionController::removePage): Call new removeProcessPool() and removeUserContentController() helpers.
(WebKit::WebExtensionController::addProcessPool): Added.
(WebKit::WebExtensionController::removeProcessPool): Added.
(WebKit::WebExtensionController::addUserContentController): Added.
(WebKit::WebExtensionController::removeUserContentController): Added.
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionMatchPatternCocoa.mm:
(WebKit::WebExtensionMatchPattern::patternsMatchAllHosts): Added.
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionURLSchemeHandlerCocoa.mm:
(WebKit::WebExtensionURLSchemeHandler::platformStartTask): Fix lambda capture to capture only this and protect task.
* Source/WebKit/UIProcess/Extensions/WebExtension.h:
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.cpp:
(WebKit::WebExtensionContext::parameters const): Remove PLATFORM(COCOA) ifdefs after review comments from Darin made me reconsider them.
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.h:
(WebKit::WebExtensionContext::addInjectedContent): Added.
(WebKit::WebExtensionContext::updateInjectedContent): Added.
* Source/WebKit/UIProcess/Extensions/WebExtensionController.cpp:
(WebKit::WebExtensionController::~WebExtensionController): Remove PLATFORM(COCOA) ifdefs after review comments from Darin made me reconsider them.
(WebKit::WebExtensionController::parameters const): Ditto.
(WebKit::WebExtensionController::allProcesses const): Added.
* Source/WebKit/UIProcess/Extensions/WebExtensionController.h:
(WebKit::WebExtensionController::allPages const): Added.
(WebKit::WebExtensionController::allUserContentControllers const): Added.
(WebKit::WebExtensionController::allProcessPools const): Added.
(WebKit::WebExtensionController::sendToAllProcesses): Use new allProcesses() method.
* Source/WebKit/UIProcess/Extensions/WebExtensionMatchPattern.h:
* Source/WebKit/UIProcess/UserContent/WebUserContentControllerProxy.h:
(WebKit::WebUserContentControllerProxy::operator== const): Added. Needed for WebExtensionController::removeUserContentController().
(WebKit::WebUserContentControllerProxy::operator!= const): Added.
* Source/WebKit/UIProcess/WebPageProxy.h:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtension.mm:
(TestWebKitAPI::TEST): Use renamed _hasStaticInjectedContentForURL: method.
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionController.mm:
(TestWebKitAPI::TEST): Added WKWebExtensionController.ContentScriptLoading test.

Canonical link: <a href="https://commits.webkit.org/257122@main">https://commits.webkit.org/257122@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2f73fc6702ff5b9667cdb6481bba52d3be734cd9

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [  ~~🧪 style~~](https://ews-build.webkit.org/#/builders/6/builds/97989 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 ios~~](https://ews-build.webkit.org/#/builders/77/builds/7204 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac~~](https://ews-build.webkit.org/#/builders/43/builds/31150 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/8/builds/107453 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/167728 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [  ~~🧪 bindings~~](https://ews-build.webkit.org/#/builders/11/builds/101928 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 ios-sim~~](https://ews-build.webkit.org/#/builders/76/builds/7670 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac-debug~~](https://ews-build.webkit.org/#/builders/71/builds/35980 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/36/builds/90484 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/104076 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [  ~~🧪 webkitperl~~](https://ews-build.webkit.org/#/builders/19/builds/103630 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/5780 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/61/builds/84592 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/90484 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/87649 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/43/builds/31150 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/36/builds/90484 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/81/builds/1181 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/43/builds/31150 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/82/builds/1162 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/43/builds/31150 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/80/builds/5997 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/61/builds/84592 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/2430 "Built successfully and passed tests") | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/79/builds/2449 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/43/builds/31150 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
<!--EWS-Status-Bubble-End-->

b-weinstein commented on Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionContext.h:
```diff
@@ -294,6 +294,14 @@ WK_CLASS_AVAILABLE(macos(WK_MAC_TBA), ios(WK_IOS_TBA))
+ @abstract Checks if the extension has script or stylesheet content that can be injected into the specified URL.
+ @param url The webpage URL to check.
+ @result Returns `YES` if the extension has content that can be injected by matching the `url` against the extension's requested match patterns.
+ @discussion The extension context will still need to be loaded and have granted website permissions for its content to actually be injected.
```
`The extension context will still need to be loaded and have granted website permissions for its content to actually be injected.`

It would be nice if this caveat was made a bit more clear in the function name. Maybe `specifiesInjectedContentForURL`? I don't feel strongly about this though - what you have is fine.

xeenon replied:
Yeah `specifiesInjectedContentForURL:` might be a good name for this. I'll see what API review says.

b-weinstein commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
```diff
@@ -1133,6 +1181,288 @@ static _WKWebExtensionContextError toAPI(WebExtensionContext::Error error)
+            // In that case we need to use the include pattern scheme and path, but with the granted pattern host.
+            RefPtr restrictedPattern = includeMatchPattern.ptr();
+            if (includeMatchPattern->matchesPattern(pattern, { WebExtensionMatchPattern::Options::IgnoreSchemes, WebExtensionMatchPattern::Options::IgnorePaths }))
+                restrictedPattern = WebExtensionMatchPattern::getOrCreate(includeMatchPattern->scheme() ?: @"*", pattern.host() ?: @"*", includeMatchPattern->path() ?: @"/*");
```
It looks like we use this structure a couple places (with defaults for missing scheme, host, and path). Could this be put into a helper method?

xeenon replied:
I moved it inside `WebExtensionMatchPattern::getOrCreate()`

xeenon committed:
Support injected content in Web Extensions.
https://bugs.webkit.org/show_bug.cgi?id=246486

Reviewed by Brian Weinstein.

* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtension.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtension.mm:
(-[_WKWebExtension _hasStaticInjectedContentForURL:]): Renamed from hasInjectedContentForURL:.
(-[_WKWebExtension hasInjectedContentForURL:]): Deleted.
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionContext.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionContext.mm:
(-[_WKWebExtensionContext hasInjectedContentForURL:]): Added.
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionController.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionPrivate.h:
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionCocoa.mm:
(WebKit::WebExtension::removeError): Fix lambda capture to capture only this.
(WebKit::WebExtension::recordError): Ditto.
(WebKit::WebExtension::resourceStringForPath): Added. Helper for string resources.
(WebKit::WebExtension::resourceDataForPath): Don't replace cached strings.
(WebKit::WebExtension::staticInjectedContents): Renamed from injectedContents().
(WebKit::WebExtension::hasStaticInjectedContentForURL): Renamed from hasInjectedContentForURL().
(WebKit::WebExtension::populateContentScriptPropertiesIfNeeded): Use m_staticInjectedContents.
(WebKit::WebExtension::allRequestedMatchPatterns): Ditto.
(WebKit::WebExtension::injectedContents): Deleted.
(WebKit::WebExtension::hasInjectedContentForURL): Deleted.
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
(WebKit::WebExtensionContext::load): Call addInjectedContent(). Adjust FIXMEs.
(WebKit::WebExtensionContext::unload): Call removeInjectedContent(). Move nil setting to end.
(WebKit::WebExtensionContext::injectedContents): Added with FIXME for dynamic scripts.
(WebKit::WebExtensionContext::hasInjectedContentForURL): Added. This is better here than WebExtension
since dynamic scripts will make the answer wrong on WebExtension but correct here.
(WebKit::WebExtensionContext::postAsyncNotification): Fix lambda capture to fix crashes.
(WebKit::WebExtensionContext::grantPermissionMatchPatterns): Call addInjectedContent().
(WebKit::WebExtensionContext::denyPermissionMatchPatterns): Call updateInjectedContent().
(WebKit::WebExtensionContext::removeGrantedPermissionMatchPatterns): Call removeInjectedContent().
(WebKit::WebExtensionContext::removeDeniedPermissionMatchPatterns): Call updateInjectedContent().
(WebKit::WebExtensionContext::loadBackgroundWebView): Fix lambda capture to capture only this.
(WebKit::WebExtensionContext::addInjectedContent): Added.
(WebKit::toImpl): Added.
(WebKit::WebExtensionContext::removeInjectedContent): Added.
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionControllerCocoa.mm:
(WebKit::WebExtensionController::load): Fix the loading order again to fix some ASSERTs.
(WebKit::WebExtensionController::addPage): Call new addProcessPool() and addUserContentController() helpers.
(WebKit::WebExtensionController::removePage): Call new removeProcessPool() and removeUserContentController() helpers.
(WebKit::WebExtensionController::addProcessPool): Added.
(WebKit::WebExtensionController::removeProcessPool): Added.
(WebKit::WebExtensionController::addUserContentController): Added.
(WebKit::WebExtensionController::removeUserContentController): Added.
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionMatchPatternCocoa.mm:
(WebKit::WebExtensionMatchPattern::patternsMatchAllHosts): Added.
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionURLSchemeHandlerCocoa.mm:
(WebKit::WebExtensionURLSchemeHandler::platformStartTask): Fix lambda capture to capture only this and protect task.
* Source/WebKit/UIProcess/Extensions/WebExtension.h:
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.cpp:
(WebKit::WebExtensionContext::parameters const): Remove PLATFORM(COCOA) ifdefs after review comments from Darin made me reconsider them.
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.h:
(WebKit::WebExtensionContext::addInjectedContent): Added.
(WebKit::WebExtensionContext::updateInjectedContent): Added.
* Source/WebKit/UIProcess/Extensions/WebExtensionController.cpp:
(WebKit::WebExtensionController::~WebExtensionController): Remove PLATFORM(COCOA) ifdefs after review comments from Darin made me reconsider them.
(WebKit::WebExtensionController::parameters const): Ditto.
(WebKit::WebExtensionController::allProcesses const): Added.
* Source/WebKit/UIProcess/Extensions/WebExtensionController.h:
(WebKit::WebExtensionController::allPages const): Added.
(WebKit::WebExtensionController::allUserContentControllers const): Added.
(WebKit::WebExtensionController::allProcessPools const): Added.
(WebKit::WebExtensionController::sendToAllProcesses): Use new allProcesses() method.
* Source/WebKit/UIProcess/Extensions/WebExtensionMatchPattern.h:
* Source/WebKit/UIProcess/UserContent/WebUserContentControllerProxy.h:
(WebKit::WebUserContentControllerProxy::operator== const): Added. Needed for WebExtensionController::removeUserContentController().
(WebKit::WebUserContentControllerProxy::operator!= const): Added.
* Source/WebKit/UIProcess/WebPageProxy.h:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtension.mm:
(TestWebKitAPI::TEST): Use renamed _hasStaticInjectedContentForURL: method.
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionController.mm:
(TestWebKitAPI::TEST): Added WKWebExtensionController.ContentScriptLoading test.

Canonical link: https://commits.webkit.org/257122@main
